File: src/mocks/handlers/ranking.ts
import { http, HttpResponse, delay } from 'msw'

let catIdCounter = 10

const categories = [
  {
    id: '1',
    name: '理论学习',
    order: 1,
    status: 'active' as const,
    description: '理论学习类作品',
    createdAt: '2026-01-10',
  },
  {
    id: '2',
    name: '党建创新',
    order: 2,
    status: 'active' as const,
    description: '党建工作创新案例',
    createdAt: '2026-01-15',
  },
  {
    id: '3',
    name: '志愿服务',
    order: 3,
    status: 'active' as const,
    description: '志愿服务活动记录',
    createdAt: '2026-02-01',
  },
  {
    id: '4',
    name: '文化风采',
    order: 4,
    status: 'inactive' as const,
    description: '文化活动展示',
    createdAt: '2026-02-20',
  },
]

const submissions = [
  {
    id: '1',
    title: '社区网格化党建创新实践',
    author: '张三',
    categoryId: '2',
    categoryName: '党建创新',
    status: 'pending' as const,
    voteCount: 0,
    submittedAt: '2026-04-15',
  },
  {
    id: '2',
    title: '青年党员志愿服务队纪实',
    author: '李四',
    categoryId: '3',
    categoryName: '志愿服务',
    status: 'approved' as const,
    voteCount: 128,
    submittedAt: '2026-04-10',
  },
  {
    id: '3',
    title: '学习强国学习心得',
    author: '王五',
    categoryId: '1',
    categoryName: '理论学习',
    status: 'approved' as const,
    voteCount: 256,
    submittedAt: '2026-03-28',
  },
  {
    id: '4',
    title: '红色诗词书法作品集',
    author: '赵六',
    categoryId: '4',
    categoryName: '文化风采',
    status: 'rejected' as const,
    voteCount: 0,
    submittedAt: '2026-04-20',
  },
  {
    id: '5',
    title: '基层党支部标准化建设经验',
    author: '孙七',
    categoryId: '2',
    categoryName: '党建创新',
    status: 'pending' as const,
    voteCount: 0,
    submittedAt: '2026-05-01',
  },
]

const styleContents = [
  {
    id: '1',
    userId: '1',
    userName: '张三',
    content: '扎根基层十余年，用实际行动践行党员初心使命...',
    status: 'published' as const,
    updatedAt: '2026-04-20',
  },
  {
    id: '2',
    userId: '2',
    userName: '李四',
    content: '带领青年志愿服务队开展帮扶活动百余次...',
    status: 'published' as const,
    updatedAt: '2026-04-18',
  },
  {
    id: '3',
    userId: '3',
    userName: '王五',
    content: '坚持理论学习，获评学习强国年度标兵...',
    status: 'draft' as const,
    updatedAt: '2026-05-02',
  },
]

export const rankingHandlers = [
  // Categories CRUD
  http.get('/api/ranking/categories', async ({ request }) => {
    await delay(300)
    const url = new URL(request.url)
    const page = +(url.searchParams.get('page') || 1)
    const pageSize = +(url.searchParams.get('pageSize') || 10)
    const start = (page - 1) * pageSize
    return HttpResponse.json({
      code: 200,
      data: { list: categories.slice(start, start + pageSize), total: categories.length, page, pageSize },
      message: 'ok',
    })
  }),

  http.post('/api/ranking/categories', async ({ request }) => {
    await delay(300)
    const body = (await request.json()) as Record<string, unknown>
    const newItem = {
      ...body,
      id: String(++catIdCounter),
      createdAt: new Date().toISOString().slice(0, 10),
    } as (typeof categories)[0]
    categories.unshift(newItem)
    return HttpResponse.json({ code: 200, data: newItem, message: '创建成功' })
  }),

  http.put('/api/ranking/categories/:id', async ({ request, params }) => {
    await delay(300)
    const body = (await request.json()) as Record<string, unknown>
    const idx = categories.findIndex((c) => c.id === params.id)
    if (idx !== -1) {
      categories[idx] = { ...categories[idx], ...body } as (typeof categories)[0]
      return HttpResponse.json({ code: 200, data: categories[idx], message: '更新成功' })
    }
    return HttpResponse.json({ code: 404, data: null, message: '不存在' }, { status: 404 })
  }),

  http.delete('/api/ranking/categories/:id', async ({ params }) => {
    await delay(200)
    const idx = categories.findIndex((c) => c.id === params.id)
    if (idx !== -1) categories.splice(idx, 1)
    return HttpResponse.json({ code: 200, data: null, message: '删除成功' })
  }),

  // Submissions
  http.get('/api/ranking/submissions', async ({ request }) => {
    await delay(300)
    const url = new URL(request.url)
    const page = +(url.searchParams.get('page') || 1)
    const pageSize = +(url.searchParams.get('pageSize') || 10)
    const status = url.searchParams.get('status') || ''
    const keyword = url.searchParams.get('keyword') || ''
    let filtered = submissions
    if (status) filtered = filtered.filter((s) => s.status === status)
    if (keyword) filtered = filtered.filter((s) => s.title.includes(keyword) || s.author.includes(keyword))
    const start = (page - 1) * pageSize
    return HttpResponse.json({
      code: 200,
      data: { list: filtered.slice(start, start + pageSize), total: filtered.length, page, pageSize },
      message: 'ok',
    })
  }),

  http.put('/api/ranking/submissions/:id/approve', async ({ params }) => {
    await delay(200)
    const item = submissions.find((s) => s.id === params.id)
    if (item) item.status = 'approved'
    return HttpResponse.json({ code: 200, data: item, message: '已通过' })
  }),

  http.put('/api/ranking/submissions/:id/reject', async ({ params }) => {
    await delay(200)
    const item = submissions.find((s) => s.id === params.id)
    if (item) item.status = 'rejected'
    return HttpResponse.json({ code: 200, data: item, message: '已驳回' })
  }),

  // Style content
  http.get('/api/ranking/style', async ({ request }) => {
    await delay(300)
    const url = new URL(request.url)
    const page = +(url.searchParams.get('page') || 1)
    const pageSize = +(url.searchParams.get('pageSize') || 10)
    const start = (page - 1) * pageSize
    return HttpResponse.json({
      code: 200,
      data: { list: styleContents.slice(start, start + pageSize), total: styleContents.length, page, pageSize },
      message: 'ok',
    })
  }),

  http.put('/api/ranking/style/:id', async ({ request, params }) => {
    await delay(300)
    const body = (await request.json()) as Record<string, unknown>
    const item = styleContents.find((s) => s.id === params.id)
    if (item) {
      Object.assign(item, body, { updatedAt: new Date().toISOString().slice(0, 10) })
      return HttpResponse.json({ code: 200, data: item, message: '更新成功' })
    }
    return HttpResponse.json({ code: 404, data: null, message: '不存在' }, { status: 404 })
  }),
]

File: backend/routers/ranking.py
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException, Query
from models.common import ApiResponse, PageResult
from models.ranking import (
    RankingCategory, CategoryCreate, CategoryUpdate,
    Submission, StyleContent, StyleUpdate,
)
from utils.security import get_current_user
from database import get_collection

router = APIRouter(prefix="/ranking", tags=["Ranking"])


def format_doc(doc, model):
    return model(**{k: v for k, v in doc.items() if k != "_id"})


# ─── Categories ───────────────────────────────────────

@router.get("/categories", response_model=ApiResponse[PageResult[RankingCategory]])
async def list_categories(
    page: int = Query(1, ge=1),
    pageSize: int = Query(10, ge=1, le=100),
    keyword: str = "",
    _current_user=Depends(get_current_user),
):
    col = get_collection("ranking_categories")
    query = {}
    if keyword:
        query["name"] = {"$regex": keyword, "$options": "i"}
    total = await col.count_documents(query)
    cursor = col.find(query).skip((page - 1) * pageSize).limit(pageSize).sort("order", 1)
    items = [format_doc(doc, RankingCategory) async for doc in cursor]
    return ApiResponse(
        code=200,
        data=PageResult(list=items, total=total, page=page, pageSize=pageSize),
        message="ok",
    )


@router.post("/categories", response_model=ApiResponse[RankingCategory])
async def create_category(body: CategoryCreate, _current_user=Depends(get_current_user)):
    col = get_collection("ranking_categories")
    last = await col.find_one({}, sort=[("id", -1)])
    new_id = str(int(last["id"]) + 1) if last else "1"
    now = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    doc = body.model_dump()
    doc["id"] = new_id
    doc["createdAt"] = now
    await col.insert_one(doc)
    return ApiResponse(code=200, data=format_doc(doc, RankingCategory), message="创建成功")


@router.put("/categories/{cat_id}", response_model=ApiResponse[RankingCategory])
async def update_category(cat_id: str, body: CategoryUpdate, _current_user=Depends(get_current_user)):
    col = get_collection("ranking_categories")
    update_data = {k: v for k, v in body.model_dump().items() if v is not None}
    result = await col.find_one_and_update({"id": cat_id}, {"$set": update_data}, return_document=True)
    if result is None:
        raise HTTPException(status_code=404, detail="不存在")
    return ApiResponse(code=200, data=format_doc(result, RankingCategory), message="更新成功")


@router.delete("/categories/{cat_id}", response_model=ApiResponse[None])
async def delete_category(cat_id: str, _current_user=Depends(get_current_user)):
    col = get_collection("ranking_categories")
    result = await col.delete_one({"id": cat_id})
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="不存在")
    return ApiResponse(code=200, data=None, message="删除成功")


# ─── Submissions ──────────────────────────────────────

@router.get("/submissions", response_model=ApiResponse[PageResult[Submission]])
async def list_submissions(
    page: int = Query(1, ge=1),
    pageSize: int = Query(10, ge=1, le=100),
    keyword: str = "",
    status: str = "",
    _current_user=Depends(get_current_user),
):
    col = get_collection("submissions")
    query = {}
    if keyword:
        query["$or"] = [
            {"title": {"$regex": keyword, "$options": "i"}},
            {"author": {"$regex": keyword, "$options": "i"}},
        ]
    if status:
        query["status"] = status

    total = await col.count_documents(query)
    cursor = col.find(query).skip((page - 1) * pageSize).limit(pageSize).sort("submittedAt", -1)
    items = [format_doc(doc, Submission) async for doc in cursor]
    return ApiResponse(
        code=200,
        data=PageResult(list=items, total=total, page=page, pageSize=pageSize),
        message="ok",
    )


@router.put("/submissions/{sub_id}/approve", response_model=ApiResponse[Submission])
async def approve_submission(sub_id: str, _current_user=Depends(get_current_user)):
    col = get_collection("submissions")
    result = await col.find_one_and_update(
        {"id": sub_id}, {"$set": {"status": "approved"}}, return_document=True
    )
    if result is None:
        raise HTTPException(status_code=404, detail="不存在")
    return ApiResponse(code=200, data=format_doc(result, Submission), message="已通过")


@router.put("/submissions/{sub_id}/reject", response_model=ApiResponse[Submission])
async def reject_submission(sub_id: str, _current_user=Depends(get_current_user)):
    col = get_collection("submissions")
    result = await col.find_one_and_update(
        {"id": sub_id}, {"$set": {"status": "rejected"}}, return_document=True
    )
    if result is None:
        raise HTTPException(status_code=404, detail="不存在")
    return ApiResponse(code=200, data=format_doc(result, Submission), message="已驳回")


# ─── Style Content ────────────────────────────────────

@router.get("/style", response_model=ApiResponse[PageResult[StyleContent]])
async def list_style_contents(
    page: int = Query(1, ge=1),
    pageSize: int = Query(10, ge=1, le=100),
    keyword: str = "",
    status: str = "",
    _current_user=Depends(get_current_user),
):
    col = get_collection("style_contents")
    query = {}
    if keyword:
        query["userName"] = {"$regex": keyword, "$options": "i"}
    if status:
        query["status"] = status
    total = await col.count_documents(query)
    cursor = col.find(query).skip((page - 1) * pageSize).limit(pageSize).sort("updatedAt", -1)
    items = [format_doc(doc, StyleContent) async for doc in cursor]
    return ApiResponse(
        code=200,
        data=PageResult(list=items, total=total, page=page, pageSize=pageSize),
        message="ok",
    )


@router.put("/style/{content_id}", response_model=ApiResponse[StyleContent])
async def update_style_content(content_id: str, body: StyleUpdate, _current_user=Depends(get_current_user)):
    col = get_collection("style_contents")
    now = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    update_data = {k: v for k, v in body.model_dump().items() if v is not None}
    update_data["updatedAt"] = now
    result = await col.find_one_and_update({"id": content_id}, {"$set": update_data}, return_document=True)
    if result is None:
        raise HTTPException(status_code=404, detail="不存在")
    return ApiResponse(code=200, data=format_doc(result, StyleContent), message="更新成功")
